File: test/java/LIM2Metrics/CouplingMetrics/CBO_alledges/CBO_alledges.java
package regtest;

class Base<T>{}

class CallsClass{
  public static void callMethod(){}
}

class AccessesAttributeClass{
  public static int attribute;
}

class HasTypeClass{}

class AttribInitClass{
  public static Integer attribInit(){
    return 0;
  }
}

class InstantiatesClass{}

class ReturnsClass{}

class ThrowsClass extends Exception{}

class CanThrowClass extends Exception{}

class UsesClass{}

class ParameterClass{}

class HasArgumentsClass{}

class MyTemplateClass<T>{}

class ClassUsesClass{}

class ConstraintClass{}

class TypeArgumentConstraintClass<T>{}

//CBO = 16
class MyClass extends Base<ClassUsesClass>{
  int selfVar;
  Integer typeClsAttrib;
  HasTypeClass attrib;
  static int attribCalls;
  TypeArgumentConstraintClass<? extends ConstraintClass> typeargVar;
  
  MyClass(){
    typeClsAttrib = AttribInitClass.attribInit();
  }
  
  void selfReference(){
    selfVar++;
  }
  
  ReturnsClass foo(){
    CallsClass.callMethod();
    AccessesAttributeClass.attribute++;
    InstantiatesClass instAttrib;
    
    Object obj = new UsesClass();
    
    return new ReturnsClass();
  }
  
  void canThrowMethod() throws CanThrowClass {
  }
  
  void throwMethod(){
    try{
      throw new ThrowsClass();
    }
    catch(ThrowsClass e){}
  }
  
  void ParameterMethod(ParameterClass param){}
  
  void methodGeneric(MyTemplateClass<HasArgumentsClass> protoParam){}

  public static void main(String[] args){}
}

File: test/cpp/LIM2Metrics/CouplingMetrics/CBO_alledges/CBO_alledges.cpp
class Base{};

class AttributeCallsClass{
public:
    static int initAttribCall(){
        return int();
    }
};

class CallsClass{
public:
    static void callMethod(){}
};

class AccessesAttributeClass{
public:
    static int attribute;
};

class HasTypeClass{};

class TypeClass{
public:
    int initAttrib;
};

class AttribInitClass{
public:
    static int attribInit(){
        return int();
    }
};

class InstantiatesClass{};

class ReturnsClass{};

class ThrowsClass{};

class CanThrowClass{};

class UsesClass{};

class ParameterClass{};

class HasArgumentsClass{};

template<class T>
class MyTemplateClass{};

template<>
class MyTemplateClass<int>{};

class ClassUsesClass{};

//CBO = 17
class MyClass : public Base{
    int selfVar;
    int typeClsAttrib;
    HasTypeClass attrib;
    static int attribCalls;
    typedef ClassUsesClass classUses;
    
    MyClass() : selfVar(), typeClsAttrib(AttribInitClass::attribInit()), attrib()
    {}
    
    void selfReference(){
        selfVar;
    }
    ReturnsClass foo(){
        CallsClass::callMethod();
        AccessesAttributeClass::attribute;
        InstantiatesClass instAttrib;
        
        TypeClass typeClsAttrib = {AttribInitClass::attribInit()};
        
        UsesClass();
        
        return ReturnsClass();
    }
    
    void throwMethod() throw(CanThrowClass) {
        throw ThrowsClass();
    }
    
    void ParameterMethod(ParameterClass param){}
    
    void methodGeneric(MyTemplateClass<HasArgumentsClass> ProtoParam, MyTemplateClass<int> specParam){}


};

int MyClass::attribCalls = AttributeCallsClass::initAttribCall();

int main(){
    return 0;
}